File: app/src/main/java/com/onemoresecret/crypto/BTCAddress.java
package com.onemoresecret.crypto;

import android.util.Log;

import androidx.annotation.NonNull;

import com.onemoresecret.Util;

import org.jetbrains.annotations.Contract;
import org.spongycastle.jce.ECNamedCurveTable;
import org.spongycastle.jce.spec.ECPrivateKeySpec;
import org.spongycastle.jce.spec.ECPublicKeySpec;

import java.math.BigInteger;
import java.security.InvalidAlgorithmParameterException;
import java.security.KeyFactory;
import java.security.KeyPairGenerator;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.Security;
import java.security.interfaces.ECPrivateKey;
import java.security.interfaces.ECPublicKey;
import java.security.spec.ECGenParameterSpec;
import java.security.spec.InvalidKeySpecException;
import java.util.Arrays;
import java.util.function.Function;

public class BTCAddress {
    private static KeyPairGenerator keyPairGenerator;

    static {
        Security.insertProviderAt(new org.spongycastle.jce.provider.BouncyCastleProvider(), 1);
        var parameterSpec = new ECGenParameterSpec("secp256k1");
        try {
            keyPairGenerator = KeyPairGenerator.getInstance("EC");
            keyPairGenerator.initialize(parameterSpec);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static final String TAG = BTCAddress.class.getSimpleName();

    private static final Function<BigInteger, byte[]> toByte32 = bi -> {
        var src = bi.toByteArray();
        /* The byte representation can be either longer than 32 bytes - because of leading zeros -
         * or shorter than 32 bytes. We must handle both cases */
        var dest = new byte[32];
        System.arraycopy(
                src,
                Math.max(0, src.length - dest.length), //skip leading zeros if longer than 32 bytes
                dest,
                Math.max(0, dest.length - src.length), //prepend with leading zeros if shorter than 32 bytes
                Math.min(src.length, dest.length));
        return dest;
    };

    private static final Function<byte[], byte[]> sha256 = bArr -> {
        try {
            return MessageDigest.getInstance("SHA-256").digest(bArr);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    };

    private static final Function<byte[], byte[]> ripeMD160 = bArr -> {
        try {
            return MessageDigest.getInstance("RipeMD160").digest(bArr);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    };

    public record ECKeyPair(ECPrivateKey privateKey, ECPublicKey publicKey) {
        @NonNull
        public BTCKeyPair toBTCKeyPair() {
            return BTCAddress.toBTCKeyPair(this);
        }
    }

    @NonNull
    @Contract("_ -> new")
    public static ECKeyPair toKeyPair(byte[] privateKey) throws NoSuchAlgorithmException, InvalidKeySpecException {
        var ecParameterSpec = ECNamedCurveTable.getParameterSpec("secp256k1");
        var s = new BigInteger(1, privateKey);

        var privateKeySpec = new ECPrivateKeySpec(s,
                new org.spongycastle.jce.spec.ECParameterSpec(
                        ecParameterSpec.getCurve(),
                        ecParameterSpec.getG(),
                        ecParameterSpec.getN(),
                        ecParameterSpec.getH()));

        var publicKeySpec = new ECPublicKeySpec(ecParameterSpec.getG().multiply(s), ecParameterSpec);
        var keyFactory = KeyFactory.getInstance("EC");
        return new ECKeyPair((ECPrivateKey) keyFactory.generatePrivate(privateKeySpec), (ECPublicKey) keyFactory.generatePublic(publicKeySpec));
    }

    @NonNull
    public static ECKeyPair newKeyPair() throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
        //generate key pair
        var keyPair = keyPairGenerator.generateKeyPair();

        return new ECKeyPair((ECPrivateKey) keyPair.getPrivate(), (ECPublicKey) keyPair.getPublic());
    }

    public record BTCKeyPair(byte[] wif, byte[] btcAddress) {
        @NonNull
        @Contract(" -> new")
        public ECKeyPair toECKeyPair() throws NoSuchAlgorithmException, InvalidKeySpecException {
            return BTCAddress.toKeyPair(toPrivateKey(wif));
        }

        public String getWifBase58() {
            return Base58.encode(wif);
        }

        public String getBtcAddressBase58() {
            return Base58.encode(btcAddress);
        }
    }

    @NonNull
    public static byte[] toBTCAddress(@NonNull ECPublicKey publicKey) {
        Log.d(TAG, "Generating bitcoin address, steps as of https://gobittest.appspot.com/Address");

        var ecPoint = publicKey.getW();
        var x = toByte32.apply(ecPoint.getAffineX());
        var y = toByte32.apply(ecPoint.getAffineY());
        var btcPublicKey = new byte[65];
        btcPublicKey[0] = 0x04;
        System.arraycopy(x, 0, btcPublicKey, 1, x.length);
        System.arraycopy(y, 0, btcPublicKey, x.length + 1, y.length);

        Log.d(TAG, "1: " + Util.byteArrayToHex(btcPublicKey, false).toUpperCase());

        //SHA-256 and RIPEMD digest
        var digest = ripeMD160.apply(
                sha256.apply(btcPublicKey));

        Log.d(TAG, "3: " + Util.byteArrayToHex(digest, false).toUpperCase());

        //prepend with version byte (0x00)
        var digestWithVersionByte = new byte[digest.length + 1];
        System.arraycopy(digest, 0, digestWithVersionByte, 1, digest.length);

        Log.d(TAG, "4: " + Util.byteArrayToHex(digestWithVersionByte, false).toUpperCase());

        //calculate checksum
        var checksum = sha256.apply(sha256.apply(digestWithVersionByte));

        Log.d(TAG, "6: " + Util.byteArrayToHex(checksum, false).toUpperCase());

        //BTC address = version byte + digest + first 4 bytes of the checksum
        var btcAddress = new byte[digestWithVersionByte.length + 4];
        System.arraycopy(digestWithVersionByte, 0, btcAddress, 0, digestWithVersionByte.length);
        //append first 4 bytes of digest
        System.arraycopy(checksum, 0, btcAddress, digestWithVersionByte.length, 4);

        Log.d(TAG, "8: " + Util.byteArrayToHex(btcAddress, false).toUpperCase());

        var btcAddressBase58 = Base58.encode(btcAddress);

        Log.d(TAG, "9: " + btcAddressBase58);

        return btcAddress;
    }

    @NonNull
    public static BTCKeyPair toBTCKeyPair(@NonNull ECKeyPair keyPair) {
        //private key
        var s = keyPair.privateKey.getS();
        var privateKey = toByte32.apply(s);

        return new BTCKeyPair(toWIF(privateKey), toBTCAddress(keyPair.publicKey));
    }

    /**
     * Convert private key to WIF format.
     *
     * @param privateKey Private Key to encode
     * @return WIF byte array. Encrypt it with {@link Base58#encode(byte[])} for the readable representation.
     */
    @NonNull
    public static byte[] toWIF(@NonNull byte[] privateKey) {
        //as of https://gobittest.appspot.com/PrivateKey

        //prepend with 0x80
        var withQualifier = new byte[privateKey.length + 1];
        withQualifier[0] = (byte) 0x80;
        System.arraycopy(privateKey, 0, withQualifier, 1, privateKey.length);

        var digest = sha256.apply(sha256.apply(withQualifier));

        var wif = new byte[withQualifier.length + 4];
        System.arraycopy(withQualifier, 0, wif, 0, withQualifier.length);
        //append first 4 bytes of digest
        System.arraycopy(digest, 0, wif, withQualifier.length, 4);

        return wif;
    }

    public static boolean validateWIF(String wifString) {
        //as of https://gobittest.appspot.com/PrivateKey

        //decode Base58
        var wif = Base58.decode(wifString);

        //drop last 4 bytes
        var withQualifier = Arrays.copyOfRange(wif, 0, wif.length - 4);

        //calculate dual sha256
        var digest = sha256.apply(sha256.apply(withQualifier));

        //compare checksums
        return Arrays.equals(
                Arrays.copyOfRange(digest, 0, 4), //first 4 bytes of the digest
                Arrays.copyOfRange(wif, wif.length - 4, wif.length) //last 4 bytes of WIF byte array
        );
    }

    /**
     * Restore private key from WIF
     *
     * @param wif Wallet Interchange Format as byte array (apply {@link Base58#decode(String)} to WIF string first)
     */

    public static byte[] toPrivateKey(byte[] wif) {
        //drop first one and last 4 bytes
        return Arrays.copyOfRange(wif, 1, wif.length - 4);
    }

}

File: app/src/main/java/com/onemoresecret/crypto/BTCAddress.kt
package com.onemoresecret.crypto

import com.onemoresecret.crypto.Base58.decode
import fr.acinq.secp256k1.Secp256k1
import org.jetbrains.annotations.Contract
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.security.SecureRandom

object BTCAddress {
    private val TAG: String = BTCAddress::class.java.simpleName
    private val secureRandom = SecureRandom()

    private fun sha256(bytes: ByteArray): ByteArray {
        try {
            return MessageDigest.getInstance("SHA-256").digest(bytes)
        } catch (e: NoSuchAlgorithmException) {
            throw RuntimeException(e)
        }
    }

    private fun requirePrivateKeyBytes(privateKey: ByteArray): ByteArray {
        require(privateKey.size == 32) { "Private key must be 32 bytes" }
        return privateKey
    }

    private fun requirePrivateKey(privateKey: ByteArray): ByteArray {
        val normalizedPrivateKey = requirePrivateKeyBytes(privateKey)
        require(Secp256k1.secKeyVerify(normalizedPrivateKey)) { "Invalid secp256k1 private key" }
        return normalizedPrivateKey
    }

    @JvmStatic
    @Contract("_ -> new")
    fun toKeyPair(privateKey: ByteArray?): ECKeyPair {
        val normalizedPrivateKey = requirePrivateKey(requireNotNull(privateKey))
        return ECKeyPair(
            normalizedPrivateKey,
            Secp256k1.pubkeyCreate(normalizedPrivateKey)
        )
    }

    @JvmStatic
    fun newKeyPair(): ECKeyPair {
        val privateKey = ByteArray(32)
        do {
            secureRandom.nextBytes(privateKey)
        } while (!Secp256k1.secKeyVerify(privateKey))

        return ECKeyPair(privateKey, Secp256k1.pubkeyCreate(privateKey))
    }

    fun toBTCAddress(publicKey: ByteArray): ByteArray {
        // https://gobittest.appspot.com/Address
        require(publicKey.size == 65 && publicKey[0] == 0x04.toByte()) {
            "Public key must be an uncompressed 65-byte secp256k1 public key"
        }

        //SHA-256 and RIPEMD digest
        val digest = Ripemd160.digest(sha256(publicKey))

        //prepend with version byte (0x00)
        val digestWithVersionByte = ByteArray(digest.size + 1)
        System.arraycopy(digest, 0, digestWithVersionByte, 1, digest.size)

        //calculate checksum
        val checksum = sha256(sha256(digestWithVersionByte))

        //BTC address = version byte + digest + first 4 bytes of the checksum
        val btcAddress = ByteArray(digestWithVersionByte.size + 4)
        System.arraycopy(digestWithVersionByte, 0, btcAddress, 0, digestWithVersionByte.size)
        //append first 4 bytes of digest
        System.arraycopy(checksum, 0, btcAddress, digestWithVersionByte.size, 4)

        return btcAddress
    }

    fun toBTCKeyPair(keyPair: ECKeyPair): BTCKeyPair {
        return BTCKeyPair(toWIF(keyPair.privateKey), toBTCAddress(keyPair.publicKey))
    }

    /**
     * Convert private key to WIF format.
     *
     * @param privateKey Private Key to encode
     * @return WIF byte array. Encrypt it with [Base58.encode] for the readable representation.
     */
    fun toWIF(privateKey: ByteArray): ByteArray {
        //as of https://gobittest.appspot.com/PrivateKey
        val normalizedPrivateKey = requirePrivateKeyBytes(privateKey)

        //prepend with 0x80

        val withQualifier = ByteArray(normalizedPrivateKey.size + 1)
        withQualifier[0] = 0x80.toByte()
        System.arraycopy(normalizedPrivateKey, 0, withQualifier, 1, normalizedPrivateKey.size)

        val digest = sha256(sha256(withQualifier))

        val wif = ByteArray(withQualifier.size + 4)
        System.arraycopy(withQualifier, 0, wif, 0, withQualifier.size)
        //append first 4 bytes of digest
        System.arraycopy(digest, 0, wif, withQualifier.size, 4)

        return wif
    }

    fun validateWIF(wifString: String): Boolean {
        //as of https://gobittest.appspot.com/PrivateKey

        //decode Base58

        val wif = decode(wifString)

        //drop last 4 bytes
        val withQualifier = wif.copyOfRange(0, wif.size - 4)

        //calculate dual sha256
        val digest = sha256(sha256(withQualifier))

        //compare checksums
        return digest.copyOfRange(0, 4).contentEquals( //first 4 bytes of the digest
            wif.copyOfRange(wif.size - 4, wif.size) //last 4 bytes of WIF byte array
        )
    }

    /**
     * Restore private key from WIF
     *
     * @param wif Wallet Interchange Format as byte array (apply [Base58.decode] to WIF string first)
     */
    @JvmStatic
    fun toPrivateKey(wif: ByteArray): ByteArray {
        //drop first one and last 4 bytes
        return requirePrivateKeyBytes(wif.copyOfRange(1, wif.size - 4))
    }

    fun sign(data: ByteArray?, privateKey: ByteArray): ByteArray {
        val messageHash = sha256(requireNotNull(data))
        return Secp256k1.compact2der(Secp256k1.sign(messageHash, requirePrivateKey(privateKey)))
    }

    fun validate(publicKey: ByteArray, data: ByteArray?, signature: ByteArray?): Boolean {
        val messageHash = sha256(requireNotNull(data))
        return Secp256k1.verify(
            Secp256k1.der2compact(requireNotNull(signature)),
            messageHash,
            publicKey
        )
    }

    data class ECKeyPair(@JvmField val privateKey: ByteArray, val publicKey: ByteArray) {
        fun toBTCKeyPair(): BTCKeyPair {
            return toBTCKeyPair(this)
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is ECKeyPair) return false

            if (!privateKey.contentEquals(other.privateKey)) return false
            if (!publicKey.contentEquals(other.publicKey)) return false

            return true
        }

        override fun hashCode(): Int {
            var result = privateKey.contentHashCode()
            result = 31 * result + publicKey.contentHashCode()
            return result
        }
    }

    @JvmRecord
    data class BTCKeyPair(@JvmField val wif: ByteArray, val btcAddress: ByteArray) {
        @Contract(" -> new")
        fun toECKeyPair(): ECKeyPair {
            return toKeyPair(toPrivateKey(wif))
        }

        val wifBase58: String
            get() = Base58.encode(wif)

        val btcAddressBase58: String
            get() = Base58.encode(btcAddress)

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is BTCKeyPair) return false

            if (!wif.contentEquals(other.wif)) return false
            if (!btcAddress.contentEquals(other.btcAddress)) return false
            if (wifBase58 != other.wifBase58) return false
            if (btcAddressBase58 != other.btcAddressBase58) return false

            return true
        }

        override fun hashCode(): Int {
            var result = wif.contentHashCode()
            result = 31 * result + btcAddress.contentHashCode()
            result = 31 * result + wifBase58.hashCode()
            result = 31 * result + btcAddressBase58.hashCode()
            return result
        }
    }
}
